File: packages/compiler/native/Cells.kt
package {{androidPackage}}

/** Explicit mutable capture cell. Does not make concurrent mutation safe. */
class LucentCell(initial: Double) {
  private val lock = Any()
  private var stored = initial

  var value: Double
    get() = synchronized(lock) { stored }
    set(value) = synchronized(lock) { stored = value }
}

File: packages/compiler/native/Cells.swift
import Foundation

/// Explicit mutable capture cell. Does not make concurrent mutation safe.
final class LucentCell: @unchecked Sendable {
  private let lock = NSLock()
  private var stored: Double

  init(_ initial: Double) {
    stored = initial
  }

  var value: Double {
    get {
      lock.lock()
      defer { lock.unlock() }
      return stored
    }
    set {
      lock.lock()
      stored = newValue
      lock.unlock()
    }
  }
}
